pkg/config/loader.go

```diff
@@ -1,56 +1,141 @@
 package config
 
 import (
+	"io"
 	"os"
 
+	"github.com/go-logr/logr"
 	"github.com/kelseyhightower/envconfig"
 	"github.com/pkg/errors"
 	"sigs.k8s.io/yaml"
 
 	"github.com/kumahq/kuma/pkg/core"
 )
 
-func Load(file string, cfg Config) error {
-	return LoadWithOption(file, cfg, false, true, true)
+type Loader struct {
+	logger logr.Logger
+	cfg    Config
+
+	strict        bool
+	includeEnv    bool
+	validate      bool
+	envVarsPrefix string
+
+	Filename string
+	Content  string
 }
 
-func LoadWithOption(file string, cfg Config, strict bool, includeEnv bool, validate bool) error {
-	if file == "" {
-		core.Log.WithName("config").Info("skipping reading config from file")
-	} else if err := loadFromFile(file, cfg, strict); err != nil {
-		return err
+func NewLoader(cfg Config) *Loader {
+	return &Loader{
+		logger: core.Log.WithName("config"),
+		cfg:    cfg,
 	}
+}
 
-	if includeEnv {
-		if err := envconfig.Process("", cfg); err != nil {
-			return err
-		}
+func (l *Loader) WithStrictParsing() *Loader {
+	l.strict = true
+	return l
+}
+
+func (l *Loader) WithEnvVarsLoading(envVarsPrefix string) *Loader {
+	l.includeEnv = true
+	l.envVarsPrefix = envVarsPrefix
+	return l
+}
+
+func (l *Loader) WithValidation() *Loader {
+	l.validate = true
+	return l
+}
+
+func (l *Loader) Load(stdin io.Reader) error {
+	switch l.Filename {
+	case "-":
+		return l.LoadReader(stdin)
+	case "":
+		return l.LoadBytes([]byte(l.Content))
+	default:
+		return l.LoadFile(l.Filename)
 	}
+}
```

Comment:
Just curious of when we would call this method? Also we expose `Filename` and `Content` properties as Public but we don't have these two parameters in the NewLoader functions.

So, is this a way to allow us to use nonstandard `Loader`? Could you please add some comments for these case conditions meaning?

Reply:
I actually changed this logic and got rid of `Filename` and `Content` properties. We are calling this function here: https://github.com/kumahq/kuma/blob/c15478c28567dc156cabbf7a6e60d1b345787ff1/app/kumactl/cmd/install/install_transparent_proxy.go#L140-L144